File: library/src/main/java/com/getstream/sdk/chat/enums/Filters.java
package com.getstream.sdk.chat.enums;

import java.util.List;

/**
 * Filters a set of static methods to build a filter object
 **/
public class Filters {

    static public FilterObject and(FilterObject... filters) {
        return new FilterObject("$and", filters);
    }

    static public FilterObject or(FilterObject... filters) {
        return new FilterObject("$or", filters);
    }

    static public FilterObject nor(FilterObject... filters) {
        return new FilterObject("$nor", filters);
    }

    static public FilterObject eq(String field, Object value) {
        return new FilterObject(field, value);
    }

    static public FilterObject ne(String field, Object value) {
        return new FilterObject(field, new FilterObject("$ne", value));
    }

    static public FilterObject greaterThan(String field, Object value) {
        return new FilterObject(field, new FilterObject("$gt", value));
    }

    static public FilterObject greaterThanEquals(String field, Object value) {
        return new FilterObject(field, new FilterObject("$gte", value));
    }

    static public FilterObject lessThan(String field, Object value) {
        return new FilterObject(field, new FilterObject("$lt", value));
    }

    static public FilterObject lessThanEquals(String field, Object value) {
        return new FilterObject(field, new FilterObject("$lte", value));
    }

    static public FilterObject in(String field, String... values) {
        return new FilterObject(field, new FilterObject("$in", values));
    }

    static public FilterObject in(String field, List values) {
        return new FilterObject(field, new FilterObject("$in", values));
    }

    static public FilterObject in(String field, Number... values) {
        return new FilterObject(field, new FilterObject("$in", values));
    }

    static public FilterObject nin(String field, String... values) {
        return new FilterObject(field, new FilterObject("$nin", values));
    }

    static public FilterObject nin(String field, List values) {
        return new FilterObject(field, new FilterObject("$nin", values));
    }

    static public FilterObject nin(String field, Number... values) {
        return new FilterObject(field, new FilterObject("$nin", values));
    }
}

File: library/src/main/java/io/getstream/chat/android/client/models/Filters.kt
package io.getstream.chat.android.client.models

import io.getstream.chat.android.client.utils.FilterObject

object Filters {

    @JvmStatic
    fun contains(vararg filters: FilterObject): FilterObject {
        return FilterObject(
            "\$contains",
            filters
        )
    }

    @JvmStatic
    fun and(vararg filters: FilterObject): FilterObject {
        return FilterObject(
            "\$and",
            filters
        )
    }

    @JvmStatic
    fun or(vararg filters: FilterObject): FilterObject {
        return FilterObject(
            "\$or",
            filters
        )
    }

    @JvmStatic
    fun nor(vararg filters: FilterObject): FilterObject {
        return FilterObject(
            "\$nor",
            filters
        )
    }

    @JvmStatic
    fun eq(field: String, value: Any): FilterObject {
        return FilterObject(field, value)
    }

    @JvmStatic
    fun ne(field: String, value: Any): FilterObject {
        return FilterObject(
            field,
            FilterObject("\$ne", value)
        )
    }

    @JvmStatic
    fun greaterThan(field: String, value: Any): FilterObject {
        return FilterObject(
            field,
            FilterObject("\$gt", value)
        )
    }

    @JvmStatic
    fun greaterThanEquals(field: String, value: Any): FilterObject {
        return FilterObject(
            field,
            FilterObject("\$gte", value)
        )
    }

    @JvmStatic
    fun lessThan(field: String, value: Any): FilterObject {
        return FilterObject(
            field,
            FilterObject("\$lt", value)
        )
    }

    @JvmStatic
    fun lessThanEquals(field: String, value: Any): FilterObject {
        return FilterObject(
            field,
            FilterObject("\$lte", value)
        )
    }

    @JvmStatic
    fun `in`(field: String, vararg values: String): FilterObject {
        return FilterObject(
            field,
            FilterObject("\$in", values)
        )
    }

    @JvmStatic
    fun `in`(field: String, values: List<*>): FilterObject {
        return FilterObject(
            field,
            FilterObject("\$in", values)
        )
    }

    @JvmStatic
    fun `in`(field: String, vararg values: Number): FilterObject {
        return FilterObject(
            field,
            FilterObject("\$in", values)
        )
    }

    @JvmStatic
    fun nin(field: String, vararg values: String): FilterObject {
        return FilterObject(
            field,
            FilterObject("\$nin", values)
        )
    }

    @JvmStatic
    fun nin(field: String, values: List<*>): FilterObject {
        return FilterObject(
            field,
            FilterObject("\$nin", values)
        )
    }

    @JvmStatic
    fun nin(field: String, vararg values: Number): FilterObject {
        return FilterObject(
            field,
            FilterObject("\$nin", values)
        )
    }
}
